services/device/geolocation/geoclue_location_provider.cc

```diff
@@ -0,0 +1,377 @@
+  // restart the provider.
+  if (!high_accuracy_requested_ && high_accuracy) {
+    if (client_state_ != kStopped) {
+      StopProvider();
+    }
+    high_accuracy_requested_ = true;
+  }
+
+  if (client_state_ != kStopped) {
+    return;
+  }
+  client_state_ = kInitializing;
+
+  dbus::ObjectProxy* proxy =
+      bus_->GetObjectProxy(kServiceName, dbus::ObjectPath(kManagerObjectPath));
+
+  dbus::MethodCall call(kManagerInterfaceName, "GetClient");
+  proxy->CallMethod(
+      &call, dbus::ObjectProxy::TIMEOUT_USE_DEFAULT,
+      base::BindOnce(&GeoClueLocationProvider::OnGetClientCompleted,
+                     weak_ptr_factory_.GetWeakPtr()));
+}
+
+void GeoClueLocationProvider::StopProvider() {
+  if (client_state_ == kStopped) {
+    return;
+  }
+
+  client_state_ = kStopped;
+
+  // Invalidate weak pointers, so we don't continue any async operations.
+  weak_ptr_factory_.InvalidateWeakPtrs();
+
+  // Stop can be called before the gclue_client_ has resolved.
+  if (!gclue_client_) {
+    return;
+  }
+
+  dbus::MethodCall stop(kClientInterfaceName, "Stop");
+  gclue_client_->CallMethod(&stop, dbus::ObjectProxy::TIMEOUT_USE_DEFAULT,
+                            base::DoNothing());
+
+  // Reset pointers to dbus objects.
+  gclue_client_.reset();
+}
+
+const mojom::Geoposition& GeoClueLocationProvider::GetPosition() {
+  return last_position_;
+}
+
+void GeoClueLocationProvider::OnPermissionGranted() {
+  permission_granted_ = true;
+  MaybeStartClient();
+}
+
+void GeoClueLocationProvider::SetPosition(const mojom::Geoposition& position) {
+  last_position_ = position;
+
+  if (client_state_ != kStarted) {
+    return;
+  }
+
+  if (last_position_.error_code == mojom::Geoposition_ErrorCode::NONE &&
+      !device::ValidateGeoposition(last_position_)) {
+    return;
+  }
+  position_update_callback_.Run(this, last_position_);
+}
+
+void GeoClueLocationProvider::OnGetClientCompleted(dbus::Response* response) {
+  if (!response) {
+    SetPosition(GetErrorPosition());
+    return;
+  }
+
+  dbus::MessageReader reader(response);
+  dbus::ObjectPath path;
+  if (!reader.PopObjectPath(&path)) {
+    SetPosition(GetErrorPosition());
+    return;
+  }
+
+  gclue_client_ = bus_->GetObjectProxy(kServiceName, path);
+  SetClientProperties();
+}
+
+void GeoClueLocationProvider::SetClientProperties() {
+  auto properties = std::make_unique<GeoClueProperties>(gclue_client_.get());
+  auto* properties_raw = properties.get();
+
+  // Wait for all properties to be set before starting the client.
+  const auto callback = base::BarrierCallback<bool>(
+      2, base::BindOnce(&GeoClueLocationProvider::OnSetClientProperties,
+                        weak_ptr_factory_.GetWeakPtr(), std::move(properties)));
+
+  properties_raw->requested_accuracy_level.Set(
+      high_accuracy_requested_ ? kGeoClueAccuracyHigh : kGeoClueAccuracyLow,
+      callback);
+
+  properties_raw->desktop_id.Set(kBraveDesktopId, callback);
```

Comment:
please try to move all dbus-related object interactions into a separate class. This way you should be able to keep all properties and other similar structures always alive and avoid all this shady pointer stuff :)

pls look at `dbus/object_manager.h`. It has some support for Properties bound to the Interface. Maybe it'll make it easier to keep all this things in one place.

another example is `UPowerObject` in `services/device/battery/battery_status_manager_linux.cc`.

Reply:
Hey @goodov, I've moved the interactions into a `GeoClueClientObject` class. However, I didn't wire up the bound properties stuff, as we're not interested in any changes to the properties
1. We only set properties on GeoClue2.Client
2. We read properties from GeoClue2.Location, but a new object every time, in response to the `LocationUpdated` signal.